Inform the user if the current theme explicitly supports view transitions with its own configuration, and add a UI control to make overriding that configuration via settings optional

felixarntz opened the pull request:
## Summary

This solves the problem that currently the settings will _always_ be applied, even if the current theme already supports view transitions explicitly. At the moment, this means the actual theme support arguments are never respected, as the plugin will always apply the settings to it.

## Relevant technical choices

* Store in a global variable whether the current theme supports view transitions with custom arguments.
* If so, show a notice to the user in the settings UI to inform them about it.
* Also show a checkbox (by default unchecked) that allows to opt in to overriding the theme support arguments with the settings, simply as an extra opt-in step and safeguard.
    * If not checked, the other settings will be effectively ignored _if_ the current theme supports view transitions with custom arguments.
    * If the theme support does not use custom arguments or if it's simply polyfilled by the plugin, this checkbox is not shown because it's irrelevant. In that case, the settings will _always_ be applied.

## Testing

To test the different behavior when using a theme with explicit support, the active theme must `add_theme_support( 'view-transitions', $args )`, with a concrete list of arguments, in other words, not just `add_theme_support( 'view-transitions' )`.

For example, add this to your theme's `functions.php`:
```php
add_action(
	'after_setup_theme',
	static function () {
		add_theme_support( 'view-transitions', array( 'default-animation' => 'slide-from-right' ) );
	}
);
```

With this added, your site should now always use the "slide from right" animation for view transitions, even if the WP Admin setting for the default transition animation is set to something else (e.g. the default "fade"). _Only_ if you activate the new checkbox to override the theme's config, the other WP Admin settings for the frontend will take effect, overriding whatever the theme defines.

<!--
For maintainers only, please make sure:

- PR has a `[Type]` label.
- PR has a plugin-specific milestone, or the `no milestone` label if it does not apply to any specific plugin.
- PR has a changelog-friendly title, or the `skip changelog` label if it should not be mentioned in the plugin's changelog.
-->

Commit:
Inform the user if the current theme explicitly supports view transitions with its own configuration, and add a UI control to make overriding that configuration via settings optional.

Commit:
Only render the override UI control if relevant.

Commit:
Merge branch 'trunk' into view-transitions/respect-theme-support

Commit:
Render new field as checkbox, reusing UI control from WP Admin view transition setting.

mukeshpanchal27 commented on plugins/view-transitions/includes/settings.php:
```diff
@@ -235,17 +252,29 @@ function plvt_apply_settings_to_theme_support(): void {
+				wp_admin_notice(
+					__( 'Your theme already supports view transitions with its own adapted configuration. The settings below will override those.', 'view-transitions' ),
+					array(
+						'type'               => 'warning',
```
Since we're simply informing the user about the setting rather than alerting them to a problem, would it make more sense to use an `info` notice instead of a `warning`?

felixarntz replied:
Fair point, will update.

felixarntz replied:
Fixed in https://github.com/WordPress/performance/pull/2037/commits/108c464d83f78633d4803f21b7d09a10664c4741

mukeshpanchal27 commented on plugins/view-transitions/includes/settings.php:
```diff
@@ -140,9 +147,14 @@ function plvt_sanitize_setting( $input ): array {
+		'enable_admin_transitions',
+	);
+	foreach ( $checkbox_options as $checkbox_option ) {
+		if ( isset( $input[ $checkbox_option ] ) ) {
```
✨

mukeshpanchal27 commented on plugins/view-transitions/includes/settings.php:
```diff
@@ -191,17 +203,22 @@ function plvt_register_setting(): void {
  */
 function plvt_apply_settings_to_theme_support(): void {
-	global $_wp_theme_features;
+	global $plvt_has_theme_support_with_args, $_wp_theme_features;
```
Missing global doc for `$plvt_has_theme_support_with_args`

felixarntz replied:
Fixed in https://github.com/WordPress/performance/pull/2037/commits/c10373379c69e8ac0a42f9b80428a17d18aa40b5

mukeshpanchal27 commented on plugins/view-transitions/includes/settings.php:
```diff
@@ -235,17 +252,29 @@ function plvt_apply_settings_to_theme_support(): void {
  * @access private
  */
 function plvt_add_setting_ui(): void {
+	global $plvt_has_theme_support_with_args;
```
Missing global doc for `$plvt_has_theme_support_with_args`

felixarntz replied:
Fixed in https://github.com/WordPress/performance/pull/2037/commits/c10373379c69e8ac0a42f9b80428a17d18aa40b5

mukeshpanchal27 commented on plugins/view-transitions/includes/settings.php:
```diff
@@ -284,14 +317,20 @@ static function (): void {
 		),
 	);
+
+	// Do not render the checkbox to override if there is no thing to override.
```
```suggestion
	// Do not render the checkbox to override if there is nothing to override.
```

felixarntz replied:
Fixed in https://github.com/WordPress/performance/pull/2037/commits/2431d7a68d6dc1149dd84389bb555cd1bdab7422

mukeshpanchal27 commented on plugins/view-transitions/includes/theme.php:
```diff
@@ -28,9 +28,16 @@
  * @access private
  */
 function plvt_polyfill_theme_support(): void {
+	global $plvt_has_theme_support_with_args, $_wp_theme_features;
```
Missing global docs

felixarntz replied:
Fixed in https://github.com/WordPress/performance/pull/2037/commits/c10373379c69e8ac0a42f9b80428a17d18aa40b5

mukeshpanchal27 reviewed (changes requested):
Thanks @felixarntz for the PR!

The admin transition currently doesn’t respect the override setting, which needs to be addressed. Other than that, I’ve left some feedback comments.

```
add_action(
	'after_setup_theme',
	static function () {
		add_theme_support( 'view-transitions', array( 'default-animation' => 'slide-from-right', 'enable_admin_transitions' => false ) );
	}
);
```

felixarntz commented:
@mukeshpanchal27 
> The admin transition currently doesn’t respect the override setting, which needs to be addressed. Other than that, I’ve left some feedback comments.

I don't think the admin transition needs to respect it, as the admin view transition behavior is independent of the theme's. So the checkbox to enable view transitions in WP Admin should always apply, regardless of what the theme does. Is that not the case currently?

mukeshpanchal27 commented:
> Is that not the case currently?

Yes, it is currently the case. However, for better clarity, if this feature is meant to be independent, we should consider mentioning that either in a notice or somewhere else in the UI so the end user is aware.

felixarntz commented:
> Yes, it is currently the case. However, for better clarity, if this feature is meant to be independent, we should consider mentioning that either in a notice or somewhere else in the UI so the end user is aware.

Would it make sense to have that checkbox in an entirely separate section on the page? That would make it clear that it's not subject to the same notice. For example, in addition to the "View Transitions" section, we could have an "Admin View Transitions" section.

mukeshpanchal27 commented:
> Would it make sense to have that checkbox in an entirely separate section on the page? That would make it clear that it's not subject to the same notice. For example, in addition to the "View Transitions" section, we could have an "Admin View Transitions" section.

+1

Commit:
Merge branch 'trunk' into view-transitions/respect-theme-support

Commit:
Use info notice instead of warning notice.

Commit:
Add missing docs for global variables.

Commit:
Fix typo.

Commit:
Show separate settings section for admin view transitions controls.

felixarntz commented:
@mukeshpanchal27 I implemented a separate settings section for WP Admin in https://github.com/WordPress/performance/pull/2037/commits/469b8f0ff40d3dd76d886446d140f3024d4f3c53, and also addressed your other feedback. This is ready for another review.
